src/LagrangianParticleTracking/calculate_particle_tendency_contributions.jl

```diff
@@ -0,0 +1,75 @@
+function calculate_particle_tendency_contributions!(particles, model, active_properties)
+    arch = model.architecture
+
+    workgroup = min(length(particles), 256)
+    worksize = length(particles)
+
+    events = []
+
+    for field in active_properties
+        property = getproperty(particles.properties, field.property)
+        tendency = getproperty(model.timestepper.Gⁿ, field.tracer)
+
+        Gp_kernal! = calculate_particle_tendency_kernel!(device(arch), workgroup, worksize)
+        Gp_event = Gp_kernal!(property, tendency, particles, model.tracers, model.grid)
+
+        push!(events, Gp_event)
+    end
+
+    wait(device(arch), MultiEvent(Tuple(events)))
+
+    return nothing
+end
+
+calculate_particle_tendency_contributions!(model) = calculate_particle_tendency_contributions!(model.particles, model)
+calculate_particle_tendency_contributions!(particles::Nothing, model) = nothing
+calculate_particle_tendency_contributions!(particles, model) = !isnothing(particles.parameters) ? (:active_properties in keys(particles.parameters) ? calculate_particle_tendency_contributions!(particles, model, particles.parameters.active_properties) : nothing) : nothing
+calculate_particle_tendency_contributions!(particles, model, active_fields::Nothing) = nothing
+
+@inline function getmask(x, y, z, grid::AbstractGrid{FT, TX, TY, TZ}, loc) where {FT, TX, TY, TZ}
+    (fi, i::Int), (fj, j::Int), (fk, k::Int) = modf.(fractional_indices(x, y, z, loc, grid))
```

Comment:
Can you write this in a few lines? I'm not sure what it does. Also do we need the annotations `i::Int`? What does that do?

Reply:
This gets the x, y, z position of the particle in terms of its position on the grid in i, j, k as a tuple `(i, j, k)` `modf` then splits each of i, j, k into its integer and fractional parts i.e. `i` and `fi`. Because the integer part is used to index an array later if I don't specify it as `Int` it doesn't work because `modf` returns them both as floats.

I'm not sure how best to clarify this but could expand it to:

```suggestion
    pos = fractional_indices(x, y, z, loc, grid)
    (fi, i::Int), (fj, j::Int), (fk, k::Int) = modf.(pos)
```

Although I'm not sure this actually clarifies very much

Reply:
Is `pos` short for "position"?

Reply:
> This gets the x, y, z position of the particle in terms of its position on the grid in i, j, k as a tuple (i, j, k) modf then splits each of i, j, k into its integer and fractional parts i.e. i and fi.

I'm a little hazy on particle advection, but I think we need this kind of algorithm also to advected particles (ie, to interpolate gridded velocities to the particle position). Can we reuse the code for that?

Reply:
Yeah sorry `pos` was position. Thats a good idea, the advection interpolates the velocity fields and does have todo the same thing:
https://github.com/CliMA/Oceananigans.jl/blob/af21542ba01afe92e7fc4de6251dec55ed5182c7/src/Fields/interpolate.jl#L122-L129

I will copy the notation they use.